### PR TITLE
Material validation

### PR DESCRIPTION
**Pull request by daredevildave**
Refactor standard material loading and initialization to standardize input.
Data that is loaded for a material is validated, migrated and cleaned so that only valid data is passed into the Material object. Invalid data is ignored.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -35,13 +35,15 @@
     "karma-chrome-launcher": "^2.2.0",
     "preprocessor": "^1.4.0",
     "sinon": "^5.0.10",
-    "uglify-js": "^3.3.26"
+    "uglify-js": "^3.3.26",
+    "local-web-server": "2.6.0"
```
Comment (vkalpias) [inline package.json]: Do you really want to add this? e.g. I use `http-server`and probably people are using their own servers.

File: src/resources/material.js [hunk 2]
```diff
@@ -169,24 +42,21 @@ Object.assign(pc, function () {
...
-            // TODO: this is a bit of a mess,
-            // Probably should create a new data block for the material
-            // and put it on the asset. This preserves originally loaded asset data
-            // and can be removed/cleared when asset is unloaded.
-            if (!data.parameters) {
-                this._createParameters(data);
+            var material = this._parser.parse(data);
+
+            // temp storage for engine-only as we need this during patching
+            if (data._engine) {
```
Comment (vkalpias) [inline src/resources/material.js]: When you say `temp storage` does this mean you're changing this?

File: src/asset/asset-reference.js [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+        this._onAssetRemove = callbacks.remove;
+    };
+
+    /**
+     * @function
+     * @name pc.AssetReference#setId
+     * @description Set the asset id which this references. Either setId or setUrl must be called to initialize an asset reference
+     * @param {Number} id The asset id to which this is a reference
+     */
+    AssetReference.prototype.setId = function (id) {
```
Comment (vkalpias) [inline src/asset/asset-reference.js]: Should this be a property?

File: src/asset/asset-reference.js [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+        this._bind();
+    };
+
+    /**
+     * @function
+     * @name pc.AssetReference#setUrl
+     * @description Set the asset url which this references. Either setId or setUrl must be called to initialize an asset reference
+     * @param {String} url The asset id to which this is a reference
+     */
+    AssetReference.prototype.setUrl = function (url) {
```
Comment (vkalpias) [inline src/asset/asset-reference.js]: Should this be a property?

File: package.json (hunk 1 above)
Comment (willeastcott) [reply package.json]: Actually, I think 'serve' is the most widely used package - I'd use that: https://www.npmjs.com/package/serve

File: src/resources/material.js (hunk 2 above)
Comment (daredevildave) [reply src/resources/material.js]: This basically how it worked before, but I changed the names slightly to be a bit more explicit.

If you load a material from a json file (i.e. engine-only) we add this _engine field to the data when it is loaded (we know we are engine on when material is requested by a URL)

Then in the open() method we copy the data loaded onto the material resource temporarily.

Finally in patch() we can assign the data from the material resource into the asset.

**Pull request merged**